experimental/delegate/mlp_aie_bf16_plugin.cpp

```diff
@@ -47,7 +49,8 @@
 // Uncomment the kernel to use
 // #define DELEGATE_KERNEL_TO_USE REF_MATMUL_DELEGATE_KERNEL
 // #define DELEGATE_KERNEL_TO_USE OPT_DELEGATE_KERNEL
-#define DELEGATE_KERNEL_TO_USE LARGE_MATMUL_DELEGATE_KERNEL
+// #define DELEGATE_KERNEL_TO_USE LARGE_MATMUL_DELEGATE_KERNEL
+#define DELEGATE_KERNEL_TO_USE MATMUL_16K_DELEGATE_KERNEL
```

Comment:
What is the advantage of using macros for compile time control flow, instead of the usual c++ `if (matmul_16k) { ... }` control flow? 

The advantage of the usual c++ control flow is that CI checks everything compiles. 

I guess ideally we would classes for this, although it is in the 'experimental' directory so not necessary.

Reply:
There is currently a lot of compile-time template-based processing going on to ensure type compatibility, automatic value conversion, and copy optimization, not to mention the types in the blob of data transferred from HAL.

When I first started the delegate, it wasn't clear which direction the design was going, or how much we really want to do before heterogeneous device support makes this all obsolete.  Now that a pattern is starting to emerge, maybe we want to enhance the delegate with the ability to select its behavior at run time, via a parameter passed from HAL to select from the catalog of kernel options built in to the delegate.  I could probably preserve the type safety via object factories.  However, we don't really go backward to older kernels, except rarely for debugging, so I'm not sure if it's worth the effort.

It would be even better if I didn't have to keep modifying the code every time we added a new kernel.  That would require reading metadata from some external source, which is doable also.  Maybe just a separate header would be good.

All effort on the delegate so far has led up to this support for the 16k matmul.  I wonder what we'll want to do with it next.  I would say that if there is a next whatever, that might be the time to make delegates easier to work with.

Reply:
That makes sense to me. Heterogenous support can't come soon enough.